### ISSUE TITLE
Improving the Quickfix Python wrapper interface

### ISSUE THREAD

**Issue by kkaasbjerg** [quickfix/quickfix#588]
Dear Quickfix community,

I have recently been looking into using Quickfix's Python interface for parsing FIX-message strings (loaded from a database, i.e., no FIX engine required) and converting them to more high-level data abstractions. During this project, I have come across several issues with the wrapper code in the Python `src/python` subdirectory which I have had to patch in order to end up with a well-functioning Python package. The two most important issues concern: 1) the build and packaging process which is lacking a `setup.py` script, and hence, cannot produce a wheel, and 2) the SWIG typemaps which not always work as desired. 

Let me go into a bit more detail with the two points here.

1. The CMake build scripts are set up to compile the Python wrapper code if the build option HAVE_PYTHON=ON is set, which produces a wrapper so file which links dynamically to the native C++ quickfix so file. To get a Python package as a wheel, I have simply "wrapped" the CMake build script in a `setup.py` file containing CMake-customized `Extension` and `build_ext` classes recycling the full CMake build. So my question is now: wouldn't it be of general interest to include such a `setup.py` in the `src/python` in order for users and package distributors to be able to build Python wheels?

2. In the SWIG typemaps, there is room for improvement: 
- The first issue I stubbled upon, was that the wrapper layer converts Python input strings to UTF-8 encoded `std::string` before calling quickfix's C++ methods. In most cases this is probably fine, but in my case, the FIX messages contain Scandinavian extended-ASCII characters in which case message validation in the `quickfix::Message` constructor fails, the reason being that the Scandinavian character occupies two bytes in UTF-8 rather than the single byte it takes up in Latin1 (the standard FIX encoding and hence the message length the `BodyLength` field refers to). There are other side effects related to this issue, but they are all fixable by converting input Python strings to Latin1 encoded `std::string` instead of UTF-8 -- Python C API methods are readily available for this.
- A second issue I came across, was that the socalled `argout` typemaps which should target function/method calls where a non-void output value is passed by non-const reference in the input list are incomplete. From the Python side, you would in most cases (when the output is not a user-defined classs) prefer to remove the output value from the list of input arguments and get the wrapper layer to instantiate it and return it. This is straight-forward to handle with a few additional arguments to the SWIG typemaps. In cases where the C++ method returns a boolean indicating if the value as been set or not, the Python way would be to return the value or None.

I hope that it is in the general interest of the Quickfix community to improve the Python interface, and that the above-mentioned issues will spark some discussion on how this could be done. I have patches available that could be submitted as PRs if there is interest.

Thank you for your time,
Kristen Kaasbjerg

**Comment by ilolis** [comment]
>  wouldn't it be of general interest to include such a setup.py in the src/python in order for users and package distributors to be able to build Python wheels?
I agree that it would be better to have the python package build files inside `src/python`. Until this is done, please note that there is a repo that contains all of the files needed to build a basic quickfix wheel file: https://github.com/quickfix/quickfix-package

**Comment by abraaoz** [comment]
The repo https://github.com/quickfix/quickfix-package is unmaintained. There are some maintained forks: https://github.com/pablodcar/quickfix-py-package and https://github.com/segfault/quickfix-py-package
